relay-event-normalization/src/normalize/mod.rs

```diff
@@ -217,50 +217,170 @@ pub struct PerformanceScoreConfig {
 }
 
 /// A mapping of AI model types (like GPT-4) to their respective costs.
-#[derive(Clone, Default, Debug, Serialize, Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub struct ModelCosts {
-    /// The version of the model cost struct
-    pub version: u16,
+///
+/// Uses internally tagged representation with version field as the tag:
+/// - V1: Array-based costs with glob pattern matching for model IDs
+/// - V2: Dictionary-based costs with exact model ID keys and granular token pricing
+/// - Unknown versions: Forward compatibility for future versions
+///
+/// Example V1 JSON:
+/// ```json
+/// {
+///   "version": 1,
+///   "costs": [
+///     {
+///       "modelId": "gpt-4*",
+///       "forCompletion": false,
+///       "costPer1kTokens": 0.03
+///     }
+///   ]
+/// }
+/// ```
+///
+/// Example V2 JSON:
+/// ```json
+/// {
+///   "version": 2,
+///   "costs": {
+///     "gpt-4": {
+///       "inputPerToken": 0.03,
+///       "outputPerToken": 0.06,
+///       "outputReasoningPerToken": 0.12,
+///       "inputCachedPerToken": 0.015
+///     }
+///   }
+/// }
+/// ```
+#[derive(Clone, Debug, Serialize)]
+#[serde(tag = "version", rename_all = "camelCase")]
+pub enum ModelCosts {
+    /// Version 1 of the model costs.
+    #[serde(rename = "1")]
+    V1 {
+        /// The mappings of model ID => cost (version 1)
+        #[serde(default, skip_serializing_if = "Vec::is_empty")]
+        costs: Vec<ModelCost>,
+    },
+    /// Version 2 of the model costs.
+    #[serde(rename = "2")]
+    V2 {
+        /// The mappings of model ID => cost as a dictionary (version 2)
+        #[serde(default, skip_serializing_if = "HashMap::is_empty")]
+        costs: HashMap<String, ModelCostV2>,
+    },
+    /// Unknown version, for forward compatibility.
+    Unknown,
+}
+
+impl<'de> serde::Deserialize<'de> for ModelCosts {
+    /// Custom implementation to handle the transitive period in the version field.
+    /// Field type was changed from number to string, and serde expects strings as tags.
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde_json::Value;
+
+        let value = Value::deserialize(deserializer)?;
+
+        // Try to extract version, handling both string and integer
+        let version = match value.get("version") {
+            Some(Value::String(s)) => s.as_str(),
+            Some(Value::Number(n)) if n.as_u64() == Some(1) => "1",
+            Some(Value::Number(n)) if n.as_u64() == Some(2) => "2",
+            _ => return Ok(ModelCosts::Unknown),
+        };
 
-    /// The mappings of model ID => cost
-    #[serde(skip_serializing_if = "Vec::is_empty")]
-    pub costs: Vec<ModelCost>,
+        match version {
+            "1" => {
+                let costs = value
+                    .get("costs")
+                    .and_then(|v| serde_json::from_value(v.clone()).ok())
+                    .unwrap_or_default();
```

Comment:
You don't need to `clone` here; you can `take` the `costs` value out:
```suggestion
                let costs = value
                    .get_mut("costs")
                    .and_then(|v| serde_json::from_value(v.take()).ok())
                    .unwrap_or_default();
```
(requires `value` to be mutable)